keps/sig-cli/2368-kubectl-default-container/README.md

```diff
@@ -0,0 +1,277 @@
+
+An example Pod yaml is like below:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+ annotations:
+    kubectl.kubernetes.io/default-container: sidecar-container
+<snip>
+```
+
+### Test Plan
+Adding unit test is enough for this feature.
+
+### Graduation Criteria
+
+
+
+#### Alpha -> Beta Graduation
+
+- At least 2 release cycles pass to gather feedback and bug reports during
+- Documentations, add it to [well-known annotations docs](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/)
+
+#### Beta -> GA Graduation
+
+- Gather feedback from developers and surveys
+- At least 2 release cycles pass to gather feedback and bug reports during
+
+#### Removing a Deprecated Flag
+
+- Announce deprecation and support policy of the existing flag in 1.21
+-- As it is not compatible with `kubectl.kubernetes.io/default-logs-container` for kubectl log. Hence, this annotation will be deprecated.
+- Deprecate the flag in 1.23.
+
+
+### Upgrade / Downgrade Strategy
+None
+
+### Version Skew Strategy
+None
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [ ] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name:
+    - Components depending on the feature gate:
+  - [x] Other
+    - Describe the mechanism:
+      - This feature is explicitly opt-in since it need user to add specified
+        annotation in pod.
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+        - No. Disabling the feature would be a client behaviour.
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? 
+        - No. It is a client behaviour only.
+
+* **Does enabling the feature change any default behavior?**
+  - No. The old behavior is notification when no default container is specified.
+    Current behavior is cover that once specified annotation is in the pod spec.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  - Yes. When an Env 'KUBE_DEFAUTL_CONTAINER_ANNOTATION=disabled', we will ignore
+    the specified annotation. This Env is only for disabling this.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  - Nothing. Just unset Env KUBE_DEFAUTL_CONTAINER_ANNOTATION to enable it.
+
+* **Are there any tests for feature enablement/disablement?**
+  - There are unit tests in `staging/src/k8s.io/kubectl/pkg/cmd/exec/` and 
+    `staging/src/k8s.io/kubectl/pkg/polymorphichelpers/` that
+    verify the behaviour.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?  
+  - None
+* **What specific metrics should inform a rollback?
+  - None
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
```

Comment:
Here, recently I've discovered that folks from PRR are asking not about the unit tests, but something more related to 'is there a test that simulates the feature gate being enabled, then disabled then enabled again?'

Maybe you can cover this on the unit tests, like 'We're going to test setting the env var and then unsetting it"

Reply:
Assuming there is no automated migration of the older annotation, this doesn't really apply since on the apiserver this is just an annotation. There is no upgrade -> downgrade -> upgrade for kubectl or other client tools, at least not in the sense we care about here.